File: examples/Delphi/Table/DB/EXDBTBU.pas
{$J+} {Writable constants}

unit ExDbTbU;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  OvcDbTbl, OvcBase, OvcEF, OvcPB, OvcPF, OvcDbPF, StdCtrls, DB, DBTables,
  OvcTCmmn, OvcTCell, OvcTCStr, OvcTCEdt, Buttons, Grids, DBGrids, OvcTable,
  OvcEditU, OvcEdit, ExtCtrls, DBCtrls, OvcTCBEF, OvcTCPic, OvcTGPns,
  OvcTCHdr, OvcTCBmp, OvcTCGly, OvcTCBox, OvcTCCBx, OvcTCIco, OvcTCNum,
  OvcTCSim, OvcSF, OvcDate;

type
  TForm1 = class(TForm)
    OvcController1: TOvcController;
    TB: TOvcDbTable;
    Table1: TTable;
    DataSource1: TDataSource;
    Table2: TTable;
    OvcTCColHead1: TOvcTCColHead;
    OvcTCString1: TOvcTCString;
    OvcTCPictureField1: TOvcTCPictureField;
    OvcTCCheckBox1: TOvcTCCheckBox;
    OvcTCComboBox1: TOvcTCComboBox;
    OvcTCSimpleField1: TOvcTCSimpleField;
    OvcTCNumericField1: TOvcTCNumericField;
    OvcTCBitMap1: TOvcTCBitMap;
    OvcTCGlyph1: TOvcTCGlyph;
    cbAllowMove: TCheckBox;
    cbAllowSize: TCheckBox;
    cbAlwaysEditing: TCheckBox;
    cbCellsPaintText: TCheckBox;
    cbEnterToTab: TCheckBox;
    cbHighlightRow: TCheckBox;
    cbIntegralHeight: TCheckBox;
    cbPageScroll: TCheckBox;
    cbShowHeader: TCheckBox;
    cbShowIndicators: TCheckBox;
    cbShowPictures: TCheckBox;
    cbStretchPictures: TCheckBox;
    cbTableActive: TCheckBox;
    cbTable1: TCheckBox;
    cbTable2: TCheckBox;
    cbUseHeaderCell: TCheckBox;
    cbUseLetters: TCheckBox;
    cbUseStrings: TCheckBox;
    cbVisible: TCheckBox;
    cbWantTabs: TCheckBox;
    cbWrapAtEnds: TCheckBox;
    sfLockedColumns: TOvcSimpleField;
    sfLeftColumn: TOvcSimpleField;
    sfActiveRow: TOvcSimpleField;
    sfActiveColumn: TOvcSimpleField;
    sfHeaderHeight: TOvcSimpleField;
    sfRowHeight: TOvcSimpleField;
    btnClose: TBitBtn;
    btnUpdate: TBitBtn;
    GroupBox1: TGroupBox;
    lblClicks: TLabel;
    procedure TBLockedCellClick(Sender: TObject; RowNum: Longint;
      ColNum: Integer);
    procedure TBUnusedAreaClick(Sender: TObject; RowNum: Longint;
      ColNum: Integer);
    procedure TBActiveCellChanged(Sender: TObject; RowNum: Longint;
      ColNum: Integer);
    procedure TBIndicatorClick(Sender: TObject; RowNum: Longint;
      ColNum: Integer);
    procedure FormCreate(Sender: TObject);
    procedure cbAllowMoveClick(Sender: TObject);
    procedure cbAllowSizeClick(Sender: TObject);
    procedure cbAlwaysEditingClick(Sender: TObject);
    procedure cbCellsPaintTextClick(Sender: TObject);
    procedure cbEnterToTabClick(Sender: TObject);
    procedure cbHighlightRowClick(Sender: TObject);
    procedure cbIntegralHeightClick(Sender: TObject);
    procedure cbPageScrollClick(Sender: TObject);
    procedure cbShowHeaderClick(Sender: TObject);
    procedure cbShowIndicatorsClick(Sender: TObject);
    procedure cbShowPicturesClick(Sender: TObject);
    procedure cbStretchPicturesClick(Sender: TObject);
    procedure cbTableActiveClick(Sender: TObject);
    procedure cbTable1Click(Sender: TObject);
    procedure cbTable2Click(Sender: TObject);
    procedure cbUseHeaderCellClick(Sender: TObject);
    procedure cbUseLettersClick(Sender: TObject);
    procedure cbUseStringsClick(Sender: TObject);
    procedure cbVisibleClick(Sender: TObject);
    procedure cbWantTabsClick(Sender: TObject);
    procedure cbWrapAtEndsClick(Sender: TObject);
    procedure sfActiveColumnExit(Sender: TObject);
    procedure sfActiveRowExit(Sender: TObject);
    procedure sfLeftColumnExit(Sender: TObject);
    procedure sfLockedColumnsExit(Sender: TObject);
    procedure sfHeaderHeightExit(Sender: TObject);
    procedure sfRowHeightExit(Sender: TObject);
    procedure FormKeyPress(Sender: TObject; var Key: Char);
    procedure btnUpdateClick(Sender: TObject);
    procedure TBLeftColumnChanged(Sender: TObject);
    procedure btnCloseClick(Sender: TObject);
  private
    { Private declarations }
  public
    { Public declarations }
    procedure UpdateStatus;
  end;

var
  Form1: TForm1;

implementation

{$R *.DFM}

procedure TForm1.TBLockedCellClick(Sender: TObject; RowNum: Longint;
  ColNum: Integer);
begin
  lblClicks.Caption := Format('Locked cell click : %d,%d', [RowNum, ColNum]);
end;

procedure TForm1.TBUnusedAreaClick(Sender: TObject; RowNum: Longint;
  ColNum: Integer);
begin
  lblClicks.Caption := Format('Unused area click : %d,%d', [RowNum, ColNum]);
end;

procedure TForm1.TBIndicatorClick(Sender: TObject; RowNum: Longint;
  ColNum: Integer);
begin
  lblClicks.Caption := Format('Indicator click : %d,%d', [RowNum, ColNum]);
end;

procedure TForm1.TBActiveCellChanged(Sender: TObject; RowNum: Longint;
  ColNum: Integer);
begin
  UpdateStatus;
end;

procedure TForm1.UpdateStatus;
begin
  cbAllowMove.Checked := (dtoAllowColumnMove in TB.Options);
  cbAllowSize.Checked := (dtoAllowColumnSize in TB.Options);
  cbAlwaysEditing.Checked := (dtoAlwaysEditing in TB.Options);
  cbCellsPaintText.Checked := (dtoCellsPaintText in TB.Options);
  cbHighlightRow.Checked := (dtoHighlightActiveRow in TB.Options);
  cbIntegralHeight.Checked := (dtoIntegralHeight in TB.Options);
  cbShowHeader.Checked := (hoShowHeader in TB.HeaderOptions);
  cbShowIndicators.Checked := (dtoShowIndicators in TB.Options);
  cbShowPictures.Checked := (dtoShowPictures in TB.Options);
  cbStretchPictures.Checked := (dtoStretch in TB.Options);
  if (TB.DataSource <> nil) and (TB.DataSource.DataSet <> nil) then
    cbTableActive.Checked := TB.DataSource.DataSet.Active
  else
    cbTableActive.Checked := False;
  cbUseHeaderCell.Checked := (hoUseHeaderCell in TB.HeaderOptions);
  cbUseLetters.Checked := (hoUseLetters in TB.HeaderOptions);
  cbUseStrings.Checked := (hoUseStrings in TB.HeaderOptions);
  cbVisible.Checked := TB.Visible;
  cbWantTabs.Checked := (dtoWantTabs in TB.Options);
  cbWrapAtEnds.Checked := (dtoWrapAtEnds in TB.Options);
  cbTable1.Checked := (TB.DataSource.DataSet = Table1);
  cbTable2.Checked := (TB.DataSource.DataSet = Table2);
  sfActiveColumn.AsInteger := TB.ActiveColumn;
  sfActiveRow.AsInteger := TB.ActiveRow;
  sfHeaderHeight.AsInteger := TB.HeaderHeight;
  sfLeftColumn.AsInteger := TB.LeftColumn;
  sfLockedColumns.AsInteger := TB.LockedColumns;
  sfRowHeight.AsInteger := TB.RowHeight;
end;

procedure TForm1.FormCreate(Sender: TObject);
begin
  UpdateStatus;
end;

procedure TForm1.FormKeyPress(Sender: TObject; var Key: Char);
begin
(*
  if Key = #13 then begin
    Key := #0;
    btnUpdate.SetFocus;
    btnUpdate.Click;
  end;
*)
end;

procedure TForm1.cbAllowMoveClick(Sender: TObject);
begin
  if cbAllowMove.Checked then
    TB.Options := TB.Options + [dtoAllowColumnMove]
  else
    TB.Options := TB.Options - [dtoAllowColumnMove];
  UpdateStatus;
end;

procedure TForm1.cbAllowSizeClick(Sender: TObject);
begin
  if cbAllowSize.Checked then
    TB.Options := TB.Options + [dtoAllowColumnSize]
  else
    TB.Options := TB.Options - [dtoAllowColumnSize];
  UpdateStatus;
end;

procedure TForm1.cbAlwaysEditingClick(Sender: TObject);
begin
  if cbAlwaysEditing.Checked then
    TB.Options := TB.Options + [dtoAlwaysEditing]
  else
    TB.Options := TB.Options - [dtoAlwaysEditing];
  UpdateStatus;
end;

procedure TForm1.cbCellsPaintTextClick(Sender: TObject);
begin
  if cbCellsPaintText.Checked then
    TB.Options := TB.Options + [dtoCellsPaintText]
  else
    TB.Options := TB.Options - [dtoCellsPaintText];
  UpdateStatus;
end;

procedure TForm1.cbEnterToTabClick(Sender: TObject);
begin
  if cbEnterToTab.Checked then
    TB.Options := TB.Options + [dtoEnterToTab]
  else
    TB.Options := TB.Options - [dtoEnterToTab];
  UpdateStatus;
end;

procedure TForm1.cbHighlightRowClick(Sender: TObject);
begin
  if cbHighlightRow.Checked then
    TB.Options := TB.Options + [dtoHighlightActiveRow]
  else
    TB.Options := TB.Options - [dtoHighlightActiveRow];
  UpdateStatus;
end;

procedure TForm1.cbIntegralHeightClick(Sender: TObject);
begin
  if cbIntegralHeight.Checked then
    TB.Options := TB.Options + [dtoIntegralHeight]
  else
    TB.Options := TB.Options - [dtoIntegralHeight];
  UpdateStatus;
end;

procedure TForm1.cbPageScrollClick(Sender: TObject);
begin
  if cbPageScroll.Checked then
    TB.Options := TB.Options + [dtoPageScroll]
  else
    TB.Options := TB.Options - [dtoPageScroll];
  UpdateStatus;
end;

procedure TForm1.cbShowHeaderClick(Sender: TObject);
begin
  if cbShowHeader.Checked then
    TB.HeaderOptions := TB.HeaderOptions + [hoShowHeader]
  else
    TB.HeaderOptions := TB.HeaderOptions - [hoShowHeader];
  UpdateStatus;
end;

procedure TForm1.cbShowIndicatorsClick(Sender: TObject);
begin
  if cbShowIndicators.Checked then
    TB.Options := TB.Options + [dtoShowIndicators]
  else
    TB.Options := TB.Options - [dtoShowIndicators];
  UpdateStatus;
end;

procedure TForm1.cbShowPicturesClick(Sender: TObject);
begin
  if cbShowPictures.Checked then
    TB.Options := TB.Options + [dtoShowPictures]
  else
    TB.Options := TB.Options - [dtoShowPictures];
  UpdateStatus;
end;

procedure TForm1.cbStretchPicturesClick(Sender: TObject);
begin
  if cbStretchPictures.Checked then
    TB.Options := TB.Options + [dtoStretch]
  else
    TB.Options := TB.Options - [dtoStretch];
  UpdateStatus;
end;

procedure TForm1.cbTableActiveClick(Sender: TObject);
begin
  if (TB.DataSource <> nil) and (TB.DataSource.DataSet <> nil) then
    TB.DataSource.DataSet.Active := cbTableActive.Checked;
  UpdateStatus;
end;

procedure TForm1.cbTable1Click(Sender: TObject);
begin
  if (TB.DataSource <> nil) and (cbTable1.Checked) then begin
    cbTable2.Checked := False;
    TB.RowHeight := 20;
    TB.DataSource.DataSet := Table1;
    TB.Columns[1].DefaultCell := OvcTCString1;
    TB.Columns[12].DefaultCell := OvcTCPictureField1;
  end;
  UpdateStatus;
end;

procedure TForm1.cbTable2Click(Sender: TObject);
begin
  if (TB.DataSource <> nil) and (cbTable2.Checked) then begin
    cbTable1.Checked := False;
    TB.RowHeight := 100;
    TB.DataSource.DataSet := Table2;
    TB.Columns[0].DefaultCell := OvcTCString1;
    TB.Columns[4].DefaultCell := OvcTCBitmap1;
    TB.Columns[4].Width := 120;
  end;
  UpdateStatus;
end;

procedure TForm1.cbUseHeaderCellClick(Sender: TObject);
begin
  if cbUseHeaderCell.Checked then
    TB.HeaderOptions := TB.HeaderOptions + [hoUseHeaderCell]
  else
    TB.HeaderOptions := TB.HeaderOptions - [hoUseHeaderCell];
  UpdateStatus;
end;

procedure TForm1.cbUseLettersClick(Sender: TObject);
begin
  if cbUseLetters .Checked then
    TB.HeaderOptions := TB.HeaderOptions + [hoUseLetters]
  else
    TB.HeaderOptions := TB.HeaderOptions - [hoUseLetters];
  cbUseLetters.Checked := hoUseLetters in TB.HeaderOptions;
  cbUseStrings.Checked := hoUseStrings in TB.HeaderOptions;
end;

procedure TForm1.cbUseStringsClick(Sender: TObject);
begin
  if cbUseStrings .Checked then
    TB.HeaderOptions := TB.HeaderOptions + [hoUseStrings]
  else
    TB.HeaderOptions := TB.HeaderOptions - [hoUseStrings];
  cbUseStrings.Checked := hoUseStrings in TB.HeaderOptions;
  cbUseLetters.Checked := hoUseLetters in TB.HeaderOptions;
end;

procedure TForm1.cbVisibleClick(Sender: TObject);
begin
  TB.Visible := cbVisible.Checked;
  UpdateStatus;
end;

procedure TForm1.cbWantTabsClick(Sender: TObject);
begin
  if cbWantTabs.Checked then
    TB.Options := TB.Options + [dtoWantTabs]
  else
    TB.Options := TB.Options - [dtoWantTabs];
  UpdateStatus;
end;

procedure TForm1.cbWrapAtEndsClick(Sender: TObject);
begin
  if cbWrapAtEnds.Checked then
    TB.Options := TB.Options + [dtoWrapAtEnds]
  else
    TB.Options := TB.Options - [dtoWrapAtEnds];
  UpdateStatus;
end;

procedure TForm1.sfActiveColumnExit(Sender: TObject);
begin
  TB.ActiveColumn := sfActiveColumn.AsInteger;
  sfActiveColumn.AsInteger := TB.ActiveColumn;
end;

procedure TForm1.sfActiveRowExit(Sender: TObject);
begin
  TB.ActiveRow := sfActiveRow.AsInteger;
  sfActiveRow.AsInteger := TB.ActiveRow;
end;

procedure TForm1.sfHeaderHeightExit(Sender: TObject);
begin
  TB.HeaderHeight := sfHeaderHeight.AsInteger;
  sfHeaderHeight.AsInteger := TB.HeaderHeight;
end;

procedure TForm1.sfLeftColumnExit(Sender: TObject);
begin
  TB.LeftColumn := sfLeftColumn.AsInteger;
  sfLeftColumn.AsInteger := TB.LeftColumn;
end;

procedure TForm1.sfLockedColumnsExit(Sender: TObject);
begin
  TB.LockedColumns := sfLockedColumns.AsInteger;
  sfLockedColumns.AsInteger := TB.LockedColumns;
end;

procedure TForm1.sfRowHeightExit(Sender: TObject);
begin
  TB.RowHeight := sfRowHeight.AsInteger;
  sfRowHeight.AsInteger := TB.RowHeight;
end;

procedure TForm1.btnUpdateClick(Sender: TObject);
begin
  UpdateStatus;
end;

procedure TForm1.TBLeftColumnChanged(Sender: TObject);
begin
  UpdateStatus;
end;

procedure TForm1.btnCloseClick(Sender: TObject);
begin
  Close;
end;

end.

File: examples/CBuildr3/ExDbTbU.cpp
//---------------------------------------------------------------------------
#include <vcl\vcl.h>
#pragma hdrstop

#include "ExDbTbU.h"
//---------------------------------------------------------------------------
#pragma link "OvcDbTbl"
#pragma link "OvcTCmmn"
#pragma link "OvcBase"
#pragma link "OvcSF"
#pragma link "OvcEF"
#pragma link "OvcTCEdt"
#pragma link "OvcTCStr"
#pragma link "OvcTCell"
#pragma link "OvcTCPic"
#pragma link "OvcTCBEF"
#pragma link "OvcTCBox"
#pragma link "OvcTCGly"
#pragma link "OvcTCBmp"
#pragma link "OvcTCCBx"
#pragma link "OvcTCSim"
#pragma link "OvcTCNum"
#pragma link "OvcTCHdr"
#pragma resource "*.dfm"
TForm1 *Form1;
//---------------------------------------------------------------------------
__fastcall TForm1::TForm1(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TForm1::TBLockedCellClick(TObject *Sender, int RowNum,
  int ColNum)
{
  lblClicks->Caption = "Locked cell click : " +
    String((int)RowNum) + ", " + String((int)ColNum);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::TBUnusedAreaClick(TObject *Sender, int RowNum,
  int ColNum)
{
  lblClicks->Caption = "Unused area click : " +
    String((int)RowNum) + ", " + String((int)ColNum);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::TBIndicatorClick(TObject *Sender, int RowNum,
  int ColNum)
{
  lblClicks->Caption = "Indicator click : " +
    String((int)RowNum) + ", " + String((int)ColNum);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::TBActiveCellChanged(TObject *Sender, int RowNum,
  int ColNum)
{
  UpdateStatus();
}
//---------------------------------------------------------------------------
void TForm1::UpdateStatus()
{
  cbAllowMove->Checked = TB->Options.Contains(dtoAllowColumnMove);
  cbAllowSize->Checked = TB->Options.Contains(dtoAllowColumnSize);
  cbAlwaysEditing->Checked = TB->Options.Contains(dtoAlwaysEditing);
  cbCellsPaintText->Checked = TB->Options.Contains(dtoCellsPaintText);
  cbHighlightRow->Checked = TB->Options.Contains(dtoHighlightActiveRow);
  cbIntegralHeight->Checked = TB->Options.Contains(dtoIntegralHeight);
  cbShowHeader->Checked = TB->HeaderOptions.Contains(hoShowHeader);
  cbShowIndicators->Checked = TB->Options.Contains(dtoShowIndicators);
  cbShowPictures->Checked = TB->Options.Contains(dtoShowPictures);
  cbStretchPictures->Checked = TB->Options.Contains(dtoStretch);
  if (TB->DataSource && TB->DataSource->DataSet)
    cbTableActive->Checked = TB->DataSource->DataSet->Active;
  else
    cbTableActive->Checked = false;
  cbUseHeaderCell->Checked = TB->HeaderOptions.Contains(hoUseHeaderCell);
  cbUseLetters->Checked = TB->HeaderOptions.Contains(hoUseLetters);
  cbUseStrings->Checked = TB->HeaderOptions.Contains(hoUseStrings);
  cbVisible->Checked = TB->Visible;
  cbWantTabs->Checked = TB->Options.Contains(dtoWantTabs);
  cbWrapAtEnds->Checked = TB->Options.Contains(dtoWrapAtEnds);
  cbTable1->Checked = (TB->DataSource->DataSet == Table1);
  cbTable2->Checked = (TB->DataSource->DataSet == Table2);
  sfActiveColumn->AsInteger = TB->ActiveColumn;
  sfActiveRow->AsInteger = TB->ActiveRow;
  sfHeaderHeight->AsInteger = TB->HeaderHeight;
  sfLeftColumn->AsInteger = TB->LeftColumn;
  sfLockedColumns->AsInteger = TB->LockedColumns;
  sfRowHeight->AsInteger = TB->RowHeight;
  TB->Repaint();
}

void __fastcall TForm1::FormCreate(TObject *Sender)
{
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::FormKeyPress(TObject *Sender, char &Key)
{
/*
  if (Key == 13) {
    Key = 0);
    btnUpdate->SetFocus();
    btnUpdate->Click();
  }
*/
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbAllowMoveClick(TObject *Sender)
{
  if (cbAllowMove->Checked)
    TB->Options << dtoAllowColumnMove;
  else
    TB->Options >> dtoAllowColumnMove;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbAllowSizeClick(TObject *Sender)
{
  if (cbAllowSize->Checked)
    TB->Options << dtoAllowColumnSize;
  else
    TB->Options >> dtoAllowColumnSize;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbAlwaysEditingClick(TObject *Sender)
{
  if (cbAlwaysEditing->Checked)
    TB->Options << dtoAlwaysEditing;
  else
    TB->Options >> dtoAlwaysEditing;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbCellsPaintTextClick(TObject *Sender)
{
  if (cbCellsPaintText->Checked)
    TB->Options << dtoCellsPaintText;
  else
    TB->Options >> dtoCellsPaintText;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbEnterToTabClick(TObject *Sender)
{
  if (cbEnterToTab->Checked)
    TB->Options << dtoEnterToTab;
  else
    TB->Options >> dtoEnterToTab;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbHighlightRowClick(TObject *Sender)
{
  if (cbHighlightRow->Checked)
    TB->Options << dtoHighlightActiveRow;
  else
    TB->Options >> dtoHighlightActiveRow;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbIntegralHeightClick(TObject *Sender)
{
  if (cbIntegralHeight->Checked)
    TB->Options << dtoIntegralHeight;
  else
    TB->Options >> dtoIntegralHeight;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbPageScrollClick(TObject *Sender)
{
  if (cbPageScroll->Checked)
    TB->Options << dtoPageScroll;
  else
    TB->Options >> dtoPageScroll;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbShowHeaderClick(TObject *Sender)
{
  if (cbShowHeader->Checked)
    TB->HeaderOptions << hoShowHeader;
  else
    TB->HeaderOptions >> hoShowHeader;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbShowIndicatorsClick(TObject *Sender)
{
  if (cbShowIndicators->Checked)
    TB->Options << dtoShowIndicators;
  else
    TB->Options >> dtoShowIndicators;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbShowPicturesClick(TObject *Sender)
{
  if (cbShowPictures->Checked)
    TB->Options << dtoShowPictures;
  else
    TB->Options >> dtoShowPictures;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbStretchPicturesClick(TObject *Sender)
{
  if (cbStretchPictures->Checked)
    TB->Options << dtoStretch;
  else
    TB->Options >> dtoStretch;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbTableActiveClick(TObject *Sender)
{
  if (TB->DataSource && TB->DataSource->DataSet)
    TB->DataSource->DataSet->Active = cbTableActive->Checked;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbTable1Click(TObject *Sender)
{
  if (TB->DataSource && cbTable1->Checked) {
    cbTable2->Checked = false;
    TB->RowHeight = 20;
    TB->DataSource->DataSet = Table1;
    TB->Columns->DefaultCell[1] = OvcTCString1;
    TB->Columns->DefaultCell[12] = OvcTCPictureField1;
  }
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbTable2Click(TObject *Sender)
{
  if (TB->DataSource && cbTable2->Checked) {
    cbTable1->Checked = false;
    TB->RowHeight = 100;
    TB->DataSource->DataSet = Table2;
    TB->Columns->DefaultCell[0] = OvcTCString1;
    TB->Columns->DefaultCell[4] = OvcTCBitMap1;
    TB->Columns->Width[4] = 120;
  }
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbUseHeaderCellClick(TObject *Sender)
{
  if (cbUseHeaderCell->Checked)
    TB->HeaderOptions << hoUseHeaderCell;
  else
    TB->HeaderOptions >> hoUseHeaderCell;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbUseLettersClick(TObject *Sender)
{
  if (cbUseLetters ->Checked)
    TB->HeaderOptions << hoUseLetters;
  else
    TB->HeaderOptions >> hoUseLetters;
  cbUseLetters->Checked = TB->HeaderOptions.Contains(hoUseLetters);
  cbUseStrings->Checked = TB->HeaderOptions.Contains(hoUseStrings);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbUseStringsClick(TObject *Sender)
{
  if (cbUseStrings->Checked)
    TB->HeaderOptions << hoUseStrings;
  else
    TB->HeaderOptions >> hoUseStrings;
  cbUseStrings->Checked = TB->HeaderOptions.Contains(hoUseStrings);
  cbUseLetters->Checked = TB->HeaderOptions.Contains(hoUseLetters);
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbVisibleClick(TObject *Sender)
{
  TB->Visible = cbVisible->Checked;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbWantTabsClick(TObject *Sender)
{
  if (cbWantTabs->Checked)
    TB->Options << dtoWantTabs;
  else
    TB->Options >> dtoWantTabs;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::cbWrapAtEndsClick(TObject *Sender)
{
  if (cbWrapAtEnds->Checked)
    TB->Options << dtoWrapAtEnds;
  else
    TB->Options >> dtoWrapAtEnds;
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::sfActiveColumnExit(TObject *Sender)
{
  TB->ActiveColumn = sfActiveColumn->AsInteger;
  sfActiveColumn->AsInteger = TB->ActiveColumn;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::sfActiveRowExit(TObject *Sender)
{
  TB->ActiveRow = sfActiveRow->AsInteger;
  sfActiveRow->AsInteger = TB->ActiveRow;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::sfHeaderHeightExit(TObject *Sender)
{
  TB->HeaderHeight = sfHeaderHeight->AsInteger;
  sfHeaderHeight->AsInteger = TB->HeaderHeight;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::sfLeftColumnExit(TObject *Sender)
{
  TB->LeftColumn = sfLeftColumn->AsInteger;
  sfLeftColumn->AsInteger = TB->LeftColumn;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::sfLockedColumnsExit(TObject *Sender)
{
  TB->LockedColumns = sfLockedColumns->AsInteger;
  sfLockedColumns->AsInteger = TB->LockedColumns;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::sfRowHeightExit(TObject *Sender)
{
  TB->RowHeight = sfRowHeight->AsInteger;
  sfRowHeight->AsInteger = TB->RowHeight;
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnUpdateClick(TObject *Sender)
{
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::TBLeftColumnChanged(TObject *Sender)
{
  UpdateStatus();
}
//---------------------------------------------------------------------------
void __fastcall TForm1::btnCloseClick(TObject *Sender)
{
  Close();    
}
//---------------------------------------------------------------------------

